### PR TITLE
Improvement of OpenCL device discovery and selection

### PR DESCRIPTION
**Pull request by voidxno**
PR for fix of OpenCL device problems described in **[issue #268](https://github.com/madMAx43v3r/mmx-node/issues/268)**

_NOTE: **Only review/feedback at this stage** (complete try that is working though)_

**Goal:** Solve problem of identical platform, or device names
**Goal:** Legacy/migrate/support existing config value logic
**Goal:** Single WebGUI selection list (all devices, all platforms)
**Goal:** Possible to find/configure OpenCL device without WebGUI
**Goal:** Keep logic simple

Logic (new):
- Enum all OpenCL devices, on all platforms, in a list, log(INFO)
- If `opencl_device` is `-1`, overrides everything, 'None' (CPU)
- If no OpenCL devices, fallback to 'None' (CPU)
- If `opencl_device_name` is defined (new logic)
  - Search list `opencl_device_name` hits, independent of platform, match if counter equals `opencl_device`
  - If not found, log(WARN), fallback to 'None' (CPU)
- If `opencl_device_name` empty (hybrid/old logic)
  - If `platform` (name) empty, define it temporarily to [0] platform
  - Search list for devices with `platform` name, match if counter equals `opencl_device`
  - If not found, log(WARN), fallback to 'None' (CPU)

Comments:
- All comments in code will be removed before any final merge/PR
- Existing .json config never changed, until selection change in WebGUI
- 'First device found' eliminated visually, still works behind the scenes
- log(INFO) of discovered OpenCL devices, makes manual .json config possible

<ins>**Questions**:</ins>
- Anything you observe/want adjusted, valid, just give feedback
- ~~I have a problem with keep-it-simple and `Node.opencl_device_selidx` value (written to .json):~~
  - ~~Want a persistent/temporary value in WebGUI, on browser reload, without permanent .json write~~
  - ~~If not, will have to duplicate 'device selection' logic from Node.cpp into settings.js (don't want to)~~
- Added a new 'Reward' separator split/card/header in WebGUI (just a suggestion)

Log (sample):
```
[Node] INFO: Found OpenCL GPU device 'NVIDIA GeForce RTX 3050' [0] (NVIDIA CUDA)
[Node] INFO: Found OpenCL GPU device 'NVIDIA GeForce GT 1030' [0] (NVIDIA CUDA)
[Node] INFO: Found OpenCL GPU device 'Intel(R) Graphics [0x7d67]' [0] (Intel(R) OpenCL Graphics)
[Node] INFO: Using OpenCL GPU device 'NVIDIA GeForce RTX 3050' [0] (NVIDIA CUDA)
```

WebGUI (sample):
```
OpenCL Device (selection list):
- None
- NVIDIA GeForce RTX 3050
- NVIDIA GeForce GT 1030
- Intel(R) Graphics [0x7d67]
```

### CODE REVIEW

**Comment by voidxno** [comment]
Ref. my `Node.opencl_device_selidx` issue above. Looked at:
- `this.set_config()` call chain (`settings.js`)

One possibility:
- Add an extra boolean param to `this.set_config()`
- Indicates if write to .json/file should be performed
- `"/config/set"` (`WebAPI.cpp`)
  - `false`: Only perform `vnx::set_config`
  - `true`: Perform both `vnx::set_config` and `vnx::write_config_file` (todays default)

Not sure if right/best way. Or correct usage of vnx framework, as implemented now.

**Comment by voidxno** [comment]
`Rev02` update:
- Implementation of boolean param to `this.set_config()` (true/false write to .json)
- Eliminated `Node.opencl_device_selidx`, by using existing `Node.opencl_device`

**Comment by madMAx43v3r** [comment]
need to resolve some conflicts now

**Comment by voidxno** [comment]
`Rev03` update:
- Resolved conflicts (i.e., redone with current mainnet-rc `master` code)

File: src/Node.cpp [hunk 1]
```diff
@@ -52,55 +52,83 @@ void Node::main()
...
-			log(INFO) << "No OpenCL platform found.";
+			else { // NOTE: No OpenCL devices found, default to -1/None
+				log(INFO) << "No OpenCL devices found";
+			}
 		}
+		else { // NOTE: Config is -1/None
+			log(INFO) << "No OpenCL device used (disabled)";
+		}
+
+		vnx::write_config("Node.opencl_device", listsel); // NOTE: 'Hijack' Node.opencl_device temporary to communicate 'OpenCL Device' selection WebGUI
```
Comment (madMAx43v3r) [inline src/Node.cpp:131]: What you mean by temporary?

File: src/Node.cpp [hunk 2]
```diff
@@ -52,55 +52,83 @@ void Node::main()
...
+
+		for(const auto idp : platforms) { // NOTE: Find all OpenCL devices, insert in list
+			const auto devices = automy::basic_opencl::get_devices(idp, CL_DEVICE_TYPE_GPU);
+			for(const auto idd : devices) {
+				int device_index = 0;
+				for(size_t i = 0; i < list.size(); ++i) {
+					if( automy::basic_opencl::get_device_name(idd) == list[i].device_name ) device_index++;
+				}
+				list.push_back({automy::basic_opencl::get_device_name(idd), device_index, automy::basic_opencl::get_platform_name(idp), idp, idd});
+				log(INFO) << "Found OpenCL GPU device '" << list[list.size() - 1].device_name << "' [" << list[list.size() - 1].device_index << "] (" << list[list.size() - 1].platform_name << ")";
```
Comment (madMAx43v3r) [inline src/Node.cpp:74]: `list[list.size() - 1]` is the same as `list.back()`

File: src/Node.cpp [hunk 3]
```diff
@@ -52,55 +52,83 @@ void Node::main()
...
+			cl_device_id device_id;
+		};
+		std::vector<device> list;
+		int listsel = -1; // NOTE: Selection index for GPU entry in list, if found (-1/None if not)
+
+		for(const auto idp : platforms) { // NOTE: Find all OpenCL devices, insert in list
+			const auto devices = automy::basic_opencl::get_devices(idp, CL_DEVICE_TYPE_GPU);
+			for(const auto idd : devices) {
+				int device_index = 0;
+				for(size_t i = 0; i < list.size(); ++i) {
```
Comment (madMAx43v3r) [inline src/Node.cpp:70]: for loops like this it's better to use C++11 syntax:
```
const auto name = automy::basic_opencl::get_device_name(idd);
for(const auto& device : list) {
  if(device.device_name == name) {
    device_index++;
  }
}
```

File: src/Node.cpp [hunk 4]
```diff
@@ -52,55 +52,83 @@ void Node::main()
 		vnx::read_config("opencl.platform", platform_name);
 
 		const auto platforms = automy::basic_opencl::get_platforms();
-		{
-			std::vector<std::string> list;
-			for(const auto id : platforms) {
-				list.push_back(automy::basic_opencl::get_platform_name(id));
+
+		struct device { // NOTE: Element for each OpenCL device found
```
Comment (madMAx43v3r) [inline src/Node.cpp:56]: to keep it less verbose, since the struct is called `device`:
```
struct device_t {
  int index;
  std::string name;
  std::string platform;     // name
  ...
}
```

File: www/web-gui/public/settings.js [hunk 5]
```diff
@@ -149,57 +145,55 @@ Vue.component('node-settings', {
...
-				this.set_config("opencl.platform", value, true);
+				this.set_config("Router.open_port", value, true, true);
 			}
 		},
 		opencl_device(value, prev) {
 			if(prev != null) {
-				this.set_config("Node.opencl_device", value, true);
+				this.set_config("opencl.platform", "", true, true); // NOTE: Config change made, force any existing platform name config to ""
+				this.set_config("Node.opencl_device", (value >= 0) ? this.opencl_device_list_index[value] : -1, true, true); // NOTE: Write real relative index config value to .json file, for device, or -1/None
+				this.set_config("Node.opencl_device", value, false, true); // NOTE: 'Hijack' value temporary for device selection WebGUI (without config file write), browser reload
```
Comment (madMAx43v3r) [inline www/web-gui/public/settings.js:160]: I get it now, you can just invent a new config variable to store something in RAM.

File: src/Node.cpp [hunk 6]
```diff
@@ -52,55 +52,83 @@ void Node::main()
...
-			}
-		} else {
-			platform = automy::basic_opencl::find_platform_by_name(platform_name);
+		std::vector<std::string> device_list_name;
+		std::vector<int> device_list_index;
+		for(size_t i = 0; i < list.size(); ++i) {
+			device_list_name.push_back(list[i].device_name);
+			device_list_index.push_back(list[i].device_index);
 		}
+		vnx::write_config("Node.device_list_name", device_list_name);
```
Comment (madMAx43v3r) [inline src/Node.cpp:84]: I think it's better to use a single `std::vector<std::pair<std::string, int>>` for this. And a single config as well.

File: www/web-gui/public/settings.js [hunk 7]
```diff
@@ -149,57 +145,55 @@ Vue.component('node-settings', {
...
-		opencl_platform(value, prev) {
-			if(prev != null) {
-				this.set_config("opencl.platform", value, true);
+				this.set_config("Router.open_port", value, true, true);
 			}
 		},
 		opencl_device(value, prev) {
 			if(prev != null) {
-				this.set_config("Node.opencl_device", value, true);
+				this.set_config("opencl.platform", "", true, true); // NOTE: Config change made, force any existing platform name config to ""
```
Comment (madMAx43v3r) [inline www/web-gui/public/settings.js:158]: let's set it to `null`, will yield and empty string as well

File: www/web-gui/public/settings.js [hunk 8]
```diff
@@ -149,57 +145,55 @@ Vue.component('node-settings', {
...
+				this.set_config("Node.opencl_device", (value >= 0) ? this.opencl_device_list_index[value] : -1, true, true); // NOTE: Write real relative index config value to .json file, for device, or -1/None
+				this.set_config("Node.opencl_device", value, false, true); // NOTE: 'Hijack' value temporary for device selection WebGUI (without config file write), browser reload
+				this.set_config("Node.opencl_device_name", (value >= 0) ? this.opencl_device_list_name[value] : "", true, true); // NOTE: Device name last, visually shown 'restart needed to apply' (WebGUI), usually (threading)
```
Comment (madMAx43v3r) [inline www/web-gui/public/settings.js:161]: same here, set to `null` in fallback

File: src/Node.cpp (hunk 4 above)
Comment (madMAx43v3r) [reply src/Node.cpp:56]: oh and always initialize primitives to some default value. `int index = 0;` in this case.

**Comment by voidxno** [comment]
`Rev04` update:
- Resolved comments from `Rev03`
  - `list.back()` instead of `list[list.size() - 1]`
  - foreach C++11 loops where possible
  - renamed struct to `device_t`, less verbose, init primitive `0`
  - new RAM config variable `Node.opencl_device_select`
  - single `std::vector<std::pair<std::string, int>>` usage
  - .json `""` config values changed to `null`

NOTE: If/when get to a result that looks workable:
- Will remove all comments in code
- Do final test of all scenarios on my test machines

File: src/Node.cpp (hunk 2 above)
Comment (voidxno) [reply src/Node.cpp]: Thx, done in `Rev04` update

File: src/Node.cpp (hunk 3 above)
Comment (voidxno) [reply src/Node.cpp]: Thx, done in `Rev04` update

File: src/Node.cpp (hunk 4 above)
Comment (voidxno) [reply src/Node.cpp]: Thx, done in `Rev04` update

File: www/web-gui/public/settings.js (hunk 5 above)
Comment (voidxno) [reply www/web-gui/public/settings.js]: Thx, done in `Rev04` update

File: src/Node.cpp (hunk 6 above)
Comment (voidxno) [reply src/Node.cpp]: Thx, done in `Rev04` update

File: www/web-gui/public/settings.js (hunk 7 above)
Comment (voidxno) [reply www/web-gui/public/settings.js]: Thx, done in `Rev04` update

File: www/web-gui/public/settings.js (hunk 8 above)
Comment (voidxno) [reply www/web-gui/public/settings.js]: Thx, done in `Rev04` update

File: src/Node.cpp [hunk 9]
```diff
@@ -52,55 +52,83 @@ void Node::main()
...
+			const auto devices = automy::basic_opencl::get_devices(idp, CL_DEVICE_TYPE_GPU);
+			for(const auto idd : devices) {
+				int device_index = 0;
+				const auto name = automy::basic_opencl::get_device_name(idd);
+				for(const auto& device : list) {
+					if(device.name == name) {
+						device_index++;
+					}
+				}
+				list.push_back({device_index, automy::basic_opencl::get_device_name(idd), automy::basic_opencl::get_platform_name(idp), idd, idp});
```
Comment (madMAx43v3r) [inline src/Node.cpp:76]: can reuse `name` here, and I would store platform name outside the loop, so you don't call `automy::basic_opencl::get_platform_name` as often.

File: src/Node.cpp [hunk 10]
```diff
@@ -52,55 +52,83 @@ void Node::main()
...
+				log(INFO) << "Found OpenCL GPU device '" << list.back().name << "' [" << list.back().index << "] (" << list.back().platform << ")";
```
Comment (madMAx43v3r) [inline src/Node.cpp:77]: you can allocate the device on stack first, then use the stack variable here and push it. this way the code is cleaner as you dont query the data from the vector.

File: src/Node.cpp [hunk 11]
```diff
@@ -52,55 +52,83 @@ void Node::main()
...
-		cl_platform_id platform = nullptr;
-		if(platform_name.empty()) {
-			if(!platforms.empty()) {
-				platform = platforms[0];
-			}
-		} else {
-			platform = automy::basic_opencl::find_platform_by_name(platform_name);
+		std::vector<std::pair<std::string, int>> device_list;
+		for(const auto& device : list) {
+			device_list.push_back(std::make_pair(device.name, device.index));
```
Comment (madMAx43v3r) [inline src/Node.cpp:83]: another C++11 trick here: `device_list.emplace_back(device.name, device.index)`

File: src/Node.cpp [hunk 12]
```diff
@@ -52,55 +52,83 @@ void Node::main()
...
+						if(platform_name == list[i].platform) {
+							devidx++;
+						}
+						if(devidx >= opencl_device) {
+							listsel = i;
+						}
+					}
+				}
+				if(listsel >= 0) { // NOTE: Found OpenCL device in list to use
+					opencl_context = automy::basic_opencl::create_context(list[listsel].platform_id, {list[listsel].device_id});
```
Comment (madMAx43v3r) [inline src/Node.cpp:111]: since you use `list[listsel]` many times, it's cleaner to setup an alias for it: `const auto& device = list[listsel];`
the only catch is you cannot modify `list` while using such a reference, but you're not doing it here so it's fine.

File: src/WebAPI.cpp [hunk 13]
```diff
@@ -1138,37 +1138,41 @@ void WebAPI::http_request_async(std::shared_ptr<const vnx::addons::HttpRequest>
 		std::lock_guard lock(g_config_mutex);
 		const auto iter_key = args.field.find("key");
 		const auto iter_value = args.field.find("value");
-		if(iter_key != args.field.end() && iter_value != args.field.end())
+		const auto iter_file = args.field.find("file"); // NOTE: Boolean option to write or not to config file (.json)
```
Comment (madMAx43v3r) [inline src/WebAPI.cpp:1141]: this param should be optional and default to `true`.

File: www/web-gui/public/settings.js [hunk 14]
```diff
@@ -30,29 +29,17 @@ Vue.component('node-settings', {
 					this.loading = false;
 					this.timelord = data.timelord ? true : false;
 					this.open_port = data["Router.open_port"] ? true : false;
-					this.opencl_platform = data["opencl.platform"] != null ? data["opencl.platform"] : "";
-					this.opencl_device = data["Node.opencl_device"] != null ? data["Node.opencl_device"] : 0;
-					this.opencl_device_list = [{name: "None", value: -1}];
+					this.opencl_device = data["Node.opencl_device_select"] != null ? data["Node.opencl_device_select"] : -1; // NOTE: Current List selection (not relative, value index in WebGUI list), from temporary config (RAM only) value, not value written to config .json (to support refresh of WebGUI browser)
```
Comment (madMAx43v3r) [inline www/web-gui/public/settings.js:32]: some JS magic to make it cleaner:
`this.opencl_device = data["Node.opencl_device_select"] ?? -1;`

File: src/WebAPI.cpp (hunk 13 above)
Comment (madMAx43v3r) [reply src/WebAPI.cpp:1141]: the name `file` is a bit misleading too, it sounds like a file name. but since you want a special case, the param should be inverted, so it's true when you want the special case. it could then be named `tmp_only` for example.

File: www/web-gui/public/settings.js [hunk 15]
```diff
@@ -63,14 +50,15 @@ Vue.component('node-settings', {
 					this.plot_dirs = data["Harvester.plot_dirs"];
 				});
 		},
-		set_config(key, value, restart) {
+		set_config(key, value, file, restart) { // NOTE: Added boolean option to write or not to config file (.json)
```
Comment (madMAx43v3r) [inline www/web-gui/public/settings.js:53]: and put the new param at the end, with a default value, so you don't have to change every call

**Comment by voidxno** [comment]
`Rev05` update:
- Resolved comments from `Rev04`
  - reused `name`, less calls to `automy::basic_opencl::get_platform_name`
  - didn't allocate device on stack, but used already established variables (looked clean/efficient)
  - used `.emplace_back` for `device_list` (nice trick)
  - did the `const auto& device = list[listsel];` (cleaner look)
  - reworked `file` to an optional `tmp_only` param (default `false`, inverted logic from before)
  - used `this.opencl_device = data["Node.opencl_device_select"] ?? -1;` (neat trick)
  - reworked `set_config()` with optional `tmp_only` param at end (default `false`, inverted logic from before)

**Please check:**
- The optional default `tmp_only` logic in `settings.js` (`set_config()`)
- The optional default `tmp_only` logic in `WebAPI.cpp` (`/config/set`)
- Basically:
  - Made both areas handle `tmp_only` optionally (independent of each other)
  - Not 100% sure/happy with `const bool tmp_only = ...` in `WebAPI.cpp`
- It all works, but it is 'save config' logic, do not want a bug there

_BTW: Thanks for patience/corrections. It's a good education in C++ :-) Starting to see elements of how you manage such an efficient (and large) codebase._

File: src/Node.cpp (hunk 9 above)
Comment (voidxno) [reply src/Node.cpp]: Thx, done in `Rev05` update

File: src/Node.cpp (hunk 10 above)
Comment (voidxno) [reply src/Node.cpp]: Thx, done in `Rev05` update

File: src/Node.cpp (hunk 11 above)
Comment (voidxno) [reply src/Node.cpp]: Thx, done in `Rev05` update

File: src/Node.cpp (hunk 12 above)
Comment (voidxno) [reply src/Node.cpp]: Thx, done in `Rev05` update

File: src/WebAPI.cpp (hunk 13 above)
Comment (voidxno) [reply src/WebAPI.cpp]: Thx, done in `Rev05` update

File: www/web-gui/public/settings.js (hunk 15 above)
Comment (voidxno) [reply www/web-gui/public/settings.js]: Thx, done in `Rev05` update

File: www/web-gui/public/settings.js (hunk 14 above)
Comment (voidxno) [reply www/web-gui/public/settings.js]: Thx, done in `Rev05` update

File: src/WebAPI.cpp [hunk 16]
```diff
@@ -1138,37 +1138,41 @@ void WebAPI::http_request_async(std::shared_ptr<const vnx::addons::HttpRequest>
...
 			const auto value = iter_value->second;
 			vnx::set_config(key, value);
 
-			std::string file;
-			vnx::optional<std::string> field;
-			{
-				const auto pos = key.find('.');
-				if(pos == std::string::npos) {
-					file = key;
+			const bool tmp_only = iter_tmp_only != args.field.end() ? (iter_tmp_only->second.is_bool() ? iter_tmp_only->second : false) : false;
```
Comment (madMAx43v3r) [inline src/WebAPI.cpp:1148]: Since `args` is a `vnx::Object` and we don't require the param to be set, this can be simplified to:
```
const bool tmp_only = args["tmp_only"];
```

File: www/web-gui/public/settings.js [hunk 17]
```diff
@@ -63,14 +50,15 @@ Vue.component('node-settings', {
...
+		set_config(key, value, restart, tmp_only) { // NOTE: Added optional boolean option to write or not to config file (.json), default 'false'
 			var req = {};
 			req.key = key;
 			req.value = value;
+			req.tmp_only = tmp_only ?? false;
 			fetch('/wapi/config/set', {body: JSON.stringify(req), method: "post"})
 				.then(response => {
 					if(response.ok) {
-						this.result = {key: req.key, value: req.value, restart};
+						this.result = {key: req.key, value: req.value, restart, tmp_only: req.tmp_only};
```
Comment (madMAx43v3r) [inline www/web-gui/public/settings.js:61]: not really needed

File: www/web-gui/public/settings.js [hunk 18]
```diff
@@ -63,14 +50,15 @@ Vue.component('node-settings', {
 					this.plot_dirs = data["Harvester.plot_dirs"];
 				});
 		},
-		set_config(key, value, restart) {
+		set_config(key, value, restart, tmp_only) { // NOTE: Added optional boolean option to write or not to config file (.json), default 'false'
```
Comment (madMAx43v3r) [inline www/web-gui/public/settings.js:53]: `set_config(key, value, restart, tmp_only = false)`

File: www/web-gui/public/settings.js [hunk 19]
```diff
@@ -30,29 +29,17 @@ Vue.component('node-settings', {
...
+					this.opencl_device_list = [{name: "None", value: -1}]; // NOTE: WebGUI list of OpenCL devices, with added -1/None
 					{
-						let list = data["opencl.device_list"];
+						let list = data["Node.opencl_device_list"]; // NOTE: List of devices from Node.cpp (pair with name and relative index)
 						if(list) {
 							let i = 0;
-							for(const name of list) {
-								this.opencl_device_list.push({name: name, value: i++});
+							for(const device of list) {
+								this.opencl_device_list_relidx.push({name: device[0], index: device[1]});
```
Comment (madMAx43v3r) [inline www/web-gui/public/settings.js:40]: Looks to me like you only need `index` from this array? no need for the object

File: www/web-gui/public/settings.js [hunk 20]
```diff
@@ -157,14 +145,12 @@ Vue.component('node-settings', {
...
-			if(prev != null) {
-				this.set_config("opencl.platform", value, true);
-			}
-		},
 		opencl_device(value, prev) {
 			if(prev != null) {
-				this.set_config("Node.opencl_device", value, true);
+				this.set_config("Node.opencl_device_select", value, true, true); // NOTE: Temporary current device selection WebGUI (no config file write), survive browser reload
+				this.set_config("opencl.platform", null, true); // NOTE: Config change made, force any existing platform name config to ""
+				this.set_config("Node.opencl_device", (value >= 0) ? this.opencl_device_list_relidx[value]["index"] : -1, true); // NOTE: Write real relative index config value to .json file, for device, or -1/None
```
Comment (madMAx43v3r) [inline www/web-gui/public/settings.js:152]:
```
const device = this.opencl_device_list_relidx[value];
```
And then get the fields via `device.index` and `device.name`.

Comment (madMAx43v3r) [inline www/web-gui/public/settings.js:152]: `this.opencl_device_list_relidx[value].index` and `this.opencl_device_list_relidx[value].name`

File: www/web-gui/public/settings.js [hunk 21]
```diff
@@ -30,29 +29,17 @@ Vue.component('node-settings', {
...
+					this.opencl_device_list_relidx = [];
+					this.opencl_device_list = [{name: "None", value: -1}]; // NOTE: WebGUI list of OpenCL devices, with added -1/None
 					{
-						let list = data["opencl.device_list"];
+						let list = data["Node.opencl_device_list"]; // NOTE: List of devices from Node.cpp (pair with name and relative index)
 						if(list) {
 							let i = 0;
-							for(const name of list) {
-								this.opencl_device_list.push({name: name, value: i++});
+							for(const device of list) {
```
Comment (madMAx43v3r) [inline www/web-gui/public/settings.js:39]: `for(const [i, device] of list.entries())`

Comment (voidxno) [reply www/web-gui/public/settings.js:39]: Don't see a way to make work. Loop **inits 2x lists**:
`this.opencl_device_list_relidx.push({name: device[0], index: device[1]});`
`this.opencl_device_list.push({name: device[0], value: i++});`
**First one:** Absolute devices found, name and relative index (used to write to config).
**Second one:** Display list, name and selection index, including None/-1 (HTML drop-down).
**Second one:** `i` in this case just to init selection index value (`i++`).

File: src/WebAPI.cpp (hunk 16 above)
Comment (voidxno) [reply src/WebAPI.cpp]: Thx, done in `Rev05` update

File: www/web-gui/public/settings.js (hunk 17 above)
Comment (voidxno) [reply www/web-gui/public/settings.js]: Thx, done in `Rev05` update

File: www/web-gui/public/settings.js (hunk 18 above)
Comment (voidxno) [reply www/web-gui/public/settings.js]: Thx, done in `Rev05` update

File: www/web-gui/public/settings.js (hunk 20 above)
Comment (voidxno) [reply www/web-gui/public/settings.js]: Thx, done in `Rev05` update

File: www/web-gui/public/settings.js (hunk 21 above)
Comment (voidxno) [reply www/web-gui/public/settings.js:39]: **Maybe alternative** (cleaner/more understandable):

```
let selidx = 0;
for(const [name, relidx] of list.entries()) {
        this.opencl_device_list_relidx.push({name: name, index: relidx});
        this.opencl_device_list.push({name: name, value: selidx++});
}
```

Comment (madMAx43v3r) [reply www/web-gui/public/settings.js:39]: oh no, I just wanted to replace the `i` with a builtin counter from the loop. `list.entries()` gives you both the value and the index in the array, so you don't need to count by yourself.

**Comment by voidxno** [comment]
`Rev07` update:
- Resolved leftover **(except)** comment from `Rev05` 
  - neat trick with `for(const [i, device] of list.entries())` for `i` count

File: www/web-gui/public/settings.js (hunk 21 above)
Comment (voidxno) [reply www/web-gui/public/settings.js]: Now I get it, neat trick 👍 done in `Rev07` update

**Comment by madMAx43v3r** [comment]
Looks good now, I tested it as well, works. But I only have one GPU to select.

**Comment by madMAx43v3r** [comment]
The settings page needs to be re-worked with tabs, but in the meantime we can merge your change.

**Comment by voidxno** [comment]
Suggestion:
- Let me make a last revision (remove all 'NOTE' comments), fix static 'Rewards' label/text
- Give me a day to retest different scenarious on my test machines (several GPU's, no GPU, so on)
- Then I'll reply when finished

**Comment by voidxno** [comment]
Testing in progress (I'll reply when finished)

**Comment by voidxno** [comment]
Testing done Linux (with latest `Rev08`):
- Machine with 2x Intel GPU / 2x NVIDIA GPU
- Machine with 1x AMD GPU
- Migrate of existing config value logic
- Identical platform, or device names
- Quick test with Windows build

Merge ok from my side.
